sql/core/src/main/scala/org/apache/spark/sql/execution/joins/OuterJoin.scala

```diff
@@ -111,52 +122,43 @@ trait HashOuterJoin {
   // iterator for performance purpose.
 
   protected[this] def leftOuterIterator(
-      key: InternalRow,
       joinedRow: JoinedRow,
       rightIter: Iterable[InternalRow],
       resultProjection: InternalRow => InternalRow): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
-      if (!key.anyNull) {
-        val temp = if (rightIter != null) {
-          rightIter.collect {
-            case r if boundCondition(joinedRow.withRight(r)) => resultProjection(joinedRow).copy()
-          }
-        } else {
-          List.empty
-        }
-        if (temp.isEmpty) {
-          resultProjection(joinedRow.withRight(rightNullRow)) :: Nil
-        } else {
-          temp
+      val temp = if (rightIter != null) {
+        rightIter.collect {
```

Comment:
we shouldn't use collect and partial function. this is really inefficient and also requires copy.